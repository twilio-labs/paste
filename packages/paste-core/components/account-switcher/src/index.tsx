export * from "./AccountSwitcher";
export * from "./AccountSwitcherGroup";
export * from "./AccountSwitcherSeparator";
export * from "./AccountSwitcherItem";
export * from "./AccountSwitcherItemRadio";
export * from "./AccountSwitcherBadge";
export * from "./useAccountSwitcher";
