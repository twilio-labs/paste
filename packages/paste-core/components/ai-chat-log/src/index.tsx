export { AIChatMessage } from "./AIChatMessage";
export type { AIChatMessageProps } from "./AIChatMessage";
export { AIChatMessageActionGroup } from "./AIChatMessageActionGroup";
export type { AIChatMessageActionGroupProps } from "./AIChatMessageActionGroup";
export { AIChatMessageBody } from "./AIChatMessageBody";
export type { AIChatMessageBodyProps } from "./AIChatMessageBody";
export { AIChatMessageAuthor } from "./AIChatMessageAuthor";
export type { AIChatMessageAuthorProps } from "./AIChatMessageAuthor";
export { AIChatMessageActionCard } from "./AIChatMessageActionCard";
export type { AIChatMessageActionCardProps } from "./AIChatMessageActionCard";
export { AIChatMessageLoading } from "./AIChatMessageLoading";
export type { AIChatMessageLoadingProps } from "./AIChatMessageLoading";

export { AIChatLog } from "./AIChatLog";
export type { AIChatLogProps } from "./AIChatLog";
export { useAIChatLogger } from "./useAIChatLogger";
export type { UseAIChatLogger, AIPartialIDChat, AIChat } from "./useAIChatLogger";
export { AIChatLogger } from "./AIChatLogger";
export type { AIChatLoggerProps } from "./AIChatLogger";

export type { AIMessageVariants } from "./AIMessageContext";
