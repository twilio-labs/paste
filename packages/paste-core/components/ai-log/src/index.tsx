export { AIChatMessage } from "./AIChatMessage";
export type { AIChatMessageProps } from "./AIChatMessage";
export { AIChatMessageMeta } from "./AIChatMessageMeta";
export type { AIChatMessageMetaProps } from "./AIChatMessageMeta";
export { AIChatMessageMetaItem } from "./AIChatMessageMetaItem";
export type { AIChatMessageMetaItemProps } from "./AIChatMessageMetaItem";
export { AIChatMessageContent } from "./AIChatMessageContent";
export type { AIChatMessageContentProps } from "./AIChatMessageContent";

export { AIChatLog } from "./AIChatLog";
export type { AIChatLogProps } from "./AIChatLog";
export { useAIChatLogger } from "./useAIChatLogger";
export type { UseAIChatLogger } from "./useAIChatLogger";
export { AIChatLogger } from "./AIChatLogger";
export type { AIChatLoggerProps } from "./AIChatLogger";
