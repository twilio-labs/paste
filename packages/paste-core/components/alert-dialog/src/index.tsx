import * as React from 'react';
import * as PropTypes from 'prop-types';
import {useUID} from '@twilio-paste/uid-library';
import {useTransition} from '@twilio-paste/animation-library';
import {Box, safelySpreadBoxProps} from '@twilio-paste/box';
import type {BoxProps} from '@twilio-paste/box';
import {ModalDialogOverlay} from '@twilio-paste/modal';
import {AlertDialogHeader} from './AlertDialogHeader';
import {AlertDialogBody} from './AlertDialogBody';
import {AlertDialogContent} from './AlertDialogContent';
import {AlertDialogFooter} from './AlertDialogFooter';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const getAnimationStates = (): any => ({
  from: {opacity: 0, transform: `scale(0.675)`},
  enter: {opacity: 1, transform: `scale(1)`},
  leave: {opacity: 0, transform: `scale(0.675)`},
  // https://www.react-spring.io/docs/hooks/api
  config: {
    mass: 0.5,
    tension: 370,
    friction: 26,
  },
});

export interface AlertDialogProps extends React.HTMLAttributes<HTMLElement>, Pick<BoxProps, 'element'> {
  children: NonNullable<React.ReactNode>;
  destructive?: boolean;
  heading: string;
  isOpen: boolean;
  onConfirm: () => void;
  onConfirmLabel: string;
  onDismiss: () => void;
  onDismissLabel: string;
}

export const AlertDialog = React.forwardRef<HTMLDivElement, AlertDialogProps>(
  (
    {
      children,
      destructive,
      element = 'ALERT_DIALOG',
      heading,
      isOpen,
      onConfirm,
      onConfirmLabel,
      onDismiss,
      onDismissLabel,
      ...props
    },
    ref
  ) => {
    const transitions = useTransition(isOpen, getAnimationStates());
    const headingID = useUID();
    const bodyID = useUID();

    return (
      <>
        {transitions(
          (styles, item) =>
            item && (
              <ModalDialogOverlay isOpen={isOpen} style={{opacity: styles.opacity}}>
                <Box
                  // @ts-expect-error Render overlay as box for customization
                  as={AlertDialogContent}
                  {...safelySpreadBoxProps(props)}
                  aria-labelledby={headingID}
                  aria-describedby={bodyID}
                  element={element}
                  ref={ref}
                  role="alertdialog"
                  style={styles}
                >
                  <AlertDialogHeader headingID={headingID} element={`${element}_HEADER`}>
                    {heading}
                  </AlertDialogHeader>
                  <AlertDialogBody bodyID={bodyID} element={`${element}_BODY`}>
                    {children}
                  </AlertDialogBody>
                  <AlertDialogFooter
                    destructive={destructive}
                    element={`${element}_FOOTER`}
                    onDismiss={onDismiss}
                    onDismissLabel={onDismissLabel}
                    onConfirm={onConfirm}
                    onConfirmLabel={onConfirmLabel}
                  />
                </Box>
              </ModalDialogOverlay>
            )
        )}
      </>
    );
  }
);

AlertDialog.displayName = 'AlertDialog';

AlertDialog.propTypes = {
  children: PropTypes.node.isRequired,
  destructive: PropTypes.bool,
  element: PropTypes.string,
  heading: PropTypes.string.isRequired,
  isOpen: PropTypes.bool.isRequired,
  onConfirm: PropTypes.func.isRequired,
  onConfirmLabel: PropTypes.string.isRequired,
  onDismiss: PropTypes.func.isRequired,
  onDismissLabel: PropTypes.string.isRequired,
};
