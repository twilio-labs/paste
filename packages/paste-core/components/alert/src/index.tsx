import * as React from 'react';
import * as PropTypes from 'prop-types';
import {ValueOf} from '@twilio-paste/types';
import {Box, safelySpreadBoxProps} from '@twilio-paste/box';
import {MediaObject, MediaFigure, MediaBody} from '@twilio-paste/media-object';
import {Button} from '@twilio-paste/button';
import {CloseIcon} from '@twilio-paste/icons/esm/CloseIcon';
import {ErrorIcon} from '@twilio-paste/icons/esm/ErrorIcon';
import {InformationIcon} from '@twilio-paste/icons/esm/InformationIcon';
import {WarningIcon} from '@twilio-paste/icons/esm/WarningIcon';

export const AlertRoles = {
  ERROR: 'alert',
  NEUTRAL: 'status',
  WARNING: 'alert',
} as const;
export const AlertVariants = {
  ERROR: 'error',
  NEUTRAL: 'neutral',
  WARNING: 'warning',
} as const;
export const AlertBackgroundColors = {
  ERROR: 'colorBackgroundErrorLightest',
  NEUTRAL: 'colorBackgroundNeutralLightest',
  WARNING: 'colorBackgroundWarningLightest',
} as const;
export const AlertBorderColors = {
  ERROR: 'colorBorderErrorLight',
  NEUTRAL: 'colorBorderNeutralLight',
  WARNING: 'colorBorderWarningLight',
} as const;

export type AlertVariants = ValueOf<typeof AlertVariants>;
export type AlertBackgroundColors = ValueOf<typeof AlertBackgroundColors>;
export type AlertBorderColors = ValueOf<typeof AlertBorderColors>;
export type AlertRoles = ValueOf<typeof AlertRoles>;

export interface AlertProps {
  id?: never;
  className?: never;
  children: NonNullable<React.ReactNode>;
  onDismiss?: () => void;
  role?: string;
  variant: AlertVariants;
}

const renderAlertIcon = (variant: AlertVariants): React.ReactElement => {
  switch (variant) {
    case AlertVariants.ERROR:
      return <ErrorIcon iconColor="colorTextErrorDark" decorative={false} title="error: " size="sizeIcon30" />;
    case AlertVariants.WARNING:
      return <WarningIcon iconColor="colorTextWarningDark" decorative={false} title="warning: " size="sizeIcon30" />;
    case AlertVariants.NEUTRAL:
    default:
      return <InformationIcon iconColor="colorTextIcon" decorative={false} title="information: " size="sizeIcon30" />;
  }
};

const Alert: React.FC<AlertProps> = ({children, onDismiss, variant, role, ...props}) => {
  return (
    <Box
      {...safelySpreadBoxProps(props)}
      backgroundColor={AlertBackgroundColors[variant.toUpperCase()]}
      borderColor={AlertBorderColors[variant.toUpperCase()]}
      borderBottomWidth="borderWidth20"
      borderBottomStyle="solid"
      paddingLeft="space70"
      paddingRight="space70"
      paddingTop="space40"
      paddingBottom="space30"
      role={role != null ? role : AlertRoles[variant.toUpperCase()]}
    >
      <MediaObject as="div">
        <MediaFigure as="div" spacing="space40">
          {renderAlertIcon(variant)}
        </MediaFigure>
        <MediaBody as="div">{children}</MediaBody>
        {onDismiss && typeof onDismiss === 'function' && (
          <MediaFigure align="end" spacing="space70">
            <Button onClick={onDismiss} variant="link" size="reset">
              <CloseIcon iconColor="colorTextIcon" decorative={false} title="dismiss this alert" size="sizeIcon30" />
            </Button>
          </MediaFigure>
        )}
      </MediaObject>
    </Box>
  );
};
Alert.displayName = 'Alert';

if (process.env.NODE_ENV === 'development') {
  Alert.propTypes = {
    children: PropTypes.node.isRequired,
    onDismiss: PropTypes.func,
    role: PropTypes.string,
    variant: PropTypes.oneOf(Object.keys(AlertVariants).map(variant => AlertVariants[variant])).isRequired,
  };
}
export {Alert};
