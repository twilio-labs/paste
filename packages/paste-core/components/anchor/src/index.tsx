import * as React from 'react';
import * as PropTypes from 'prop-types';
import {compose, layout, space} from 'styled-system';
import styled from '@emotion/styled';
import css from '@styled-system/css';
import {LayoutProps, SpaceProps} from '@twilio-paste/style-props';

export type AnchorTargets = '_self' | '_blank' | '_parent' | '_top';
export type AnchorTabIndexes = 0 | -1;

export interface AnchorProps extends React.AnchorHTMLAttributes<HTMLAnchorElement>, LayoutProps, SpaceProps {
  children: NonNullable<React.ReactNode>;
  href: string;
  rel?: string;
  tabIndex?: AnchorTabIndexes;
  target?: AnchorTargets;
  id?: never;
  className?: never;
}

const EXTERNAL_URL_REGEX = /^(https?:)[^\s]*$/;
const EXTERNAL_TARGET_DEFAULT = '_blank';
const EXTERNAL_REL_DEFAULT = 'noreferrer noopener';
const isExternalUrl = (url: string): boolean => EXTERNAL_URL_REGEX.test(url);

export const secureExternalLink = (href: string): {} | undefined => {
  if (!isExternalUrl(href)) return undefined;
  return {
    rel: EXTERNAL_REL_DEFAULT,
    target: EXTERNAL_TARGET_DEFAULT,
  };
};

// eslint-disable-next-line emotion/syntax-preference
const StyledAnchor = styled.a(
  css({
    color: 'colorTextLink',
    textDecoration: 'underline',
    outline: 'none',

    '&:hover': {
      color: 'colorTextLinkDarker',
      textDecoration: 'none',
    },

    '&:focus, &:active': {
      boxShadow: 'shadowFocus',
      color: 'colorTextLinkDarker',
      textDecoration: 'none',
    },
  }),
  compose(
    space,
    layout
  )
);

const Anchor: React.FC<AnchorProps> = props => (
  <StyledAnchor {...secureExternalLink(props.href)} {...props}>
    {props.children}
  </StyledAnchor>
);
Anchor.displayName = 'Anchor';

if (process.env.NODE_ENV === 'development') {
  Anchor.propTypes = {
    children: PropTypes.node.isRequired,
    href: PropTypes.string.isRequired,
    rel: PropTypes.string,
    tabIndex: PropTypes.oneOf([0, -1]),
    target: PropTypes.oneOf(['_self', '_blank', '_parent', '_top']),
  };
}
export {Anchor};
