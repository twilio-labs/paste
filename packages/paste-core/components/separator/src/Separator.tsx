import { Box, safelySpreadBoxProps } from "@twilio-paste/box";
import type { BoxProps } from "@twilio-paste/box";
import type { Space } from "@twilio-paste/style-props";
import type { HTMLPasteProps } from "@twilio-paste/types";
import * as React from "react";

type Orientation = "horizontal" | "vertical";

interface SeparatorProps extends HTMLPasteProps<"hr"> {
  /**
   * Overrides the default element name to apply unique styles with the Customization Provider
   *
   * @default "SEPARATOR"
   * @type {BoxProps["element"]}
   * @memberof SeparatorProps
   */
  element?: BoxProps["element"];
  id?: never;
  /**
   * Separator direction
   *
   * @type {Orientation}
   * @memberof SeparatorProps
   */
  orientation: Orientation;
  /**
   * Space left and right of the separator when orientation is vertical
   *
   * @type {Space}
   * @memberof SeparatorProps
   */
  horizontalSpacing?: Space;
  /**
   * Space top and bottom of the separator when orientation is horizontal
   *
   * @type {Space}
   * @memberof SeparatorProps
   */
  verticalSpacing?: Space;
}

const Separator = React.forwardRef<HTMLHRElement, SeparatorProps>(
  ({ element = "SEPARATOR", orientation, horizontalSpacing, verticalSpacing, ...props }, ref) => {
    return (
      <Box
        {...safelySpreadBoxProps(props)}
        aria-orientation={orientation}
        element={element}
        margin="space0"
        marginBottom={verticalSpacing}
        marginLeft={horizontalSpacing}
        marginRight={horizontalSpacing}
        marginTop={verticalSpacing}
        as="hr"
        borderWidth="borderWidth0"
        borderColor="colorBorderWeaker"
        borderStyle="solid"
        borderBottomWidth={orientation === "horizontal" ? "borderWidth10" : null}
        borderLeftWidth={orientation === "vertical" ? "borderWidth10" : null}
        width={orientation === "horizontal" ? "auto" : null}
        height={orientation === "vertical" ? "auto" : null}
        ref={ref}
      />
    );
  },
);

Separator.displayName = "Separator";

export { Separator };
export type { SeparatorProps };
