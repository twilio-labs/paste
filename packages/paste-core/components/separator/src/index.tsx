export { Separator } from "./Separator";
export type { SeparatorProps } from "./Separator";
