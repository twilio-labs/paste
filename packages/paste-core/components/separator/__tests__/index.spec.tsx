import * as React from 'react';
import {render} from 'react-dom';
import renderer from 'react-test-renderer';
import {render as testRender} from '@testing-library/react';
import {Theme} from '@twilio-paste/theme';
import axe from '../../../../../.jest/axe-helper';
import {Separator} from '../src';

describe('Separator', () => {
  it('should render', () => {
    const {getByRole} = testRender(<Separator orientation="horizontal" />);
    expect(getByRole('separator')).not.toBeNull();
  });
  it('should render set horizontal orientation', () => {
    const {getByRole} = testRender(<Separator orientation="horizontal" />);
    expect(getByRole('separator').getAttribute('aria-orientation')).toBe('horizontal');
  });
  it('should render set vertical orientation', () => {
    const {getByRole} = testRender(<Separator orientation="vertical" />);
    expect(getByRole('separator').getAttribute('aria-orientation')).toBe('vertical');
  });
  it('should set horizontal margins', (): void => {
    const tree = renderer
      .create(
        <Theme.Provider>
          <Separator orientation="horizontal" horizontalSpacing="space50" />
        </Theme.Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should set vertical margins', (): void => {
    const tree = renderer
      .create(
        <Theme.Provider>
          <Separator orientation="vertical" verticalSpacing="space100" />
        </Theme.Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should set responsive horizontal margins', (): void => {
    const tree = renderer
      .create(
        <Theme.Provider>
          <Separator orientation="horizontal" horizontalSpacing={['space20', 'space100', 'space190']} />
        </Theme.Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should set responsive vertical margins', (): void => {
    const tree = renderer
      .create(
        <Theme.Provider>
          <Separator orientation="vertical" verticalSpacing={['space20', 'space100', 'space190']} />
        </Theme.Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('Should have no accessibility violations', async () => {
    const container = document.createElement('div');
    document.body.append(container);
    render(
      <>
        <Separator orientation="horizontal" />
        <Separator orientation="vertical" />
      </>,
      container
    );
    const results = await axe(document.body);
    expect(results).toHaveNoViolations();
  });
});
