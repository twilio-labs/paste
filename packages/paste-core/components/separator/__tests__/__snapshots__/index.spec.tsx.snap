// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Separator it should set horizontal margins 1`] = `
.emotion-1 {
  color: rgb(18,28,45);
  font-size: 0.875rem;
  font-family: Inter,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen-Sans,Ubuntu,Cantarell,'Helvetica Neue',sans-serif;
  line-height: 1.5rem;
  font-weight: 400;
}

.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  margin-left: 1rem;
  margin-right: 1rem;
  border-width: 0;
  border-color: rgb(225,227,234);
  border-style: solid;
  border-bottom-width: 2px;
  width: auto;
}

<div
  className="emotion-1"
>
  <hr
    aria-orientation="horizontal"
    className="emotion-0"
    height={null}
    width="auto"
  />
</div>
`;

exports[`Separator it should set responsive horizontal margins 1`] = `
.emotion-1 {
  color: rgb(18,28,45);
  font-size: 0.875rem;
  font-family: Inter,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen-Sans,Ubuntu,Cantarell,'Helvetica Neue',sans-serif;
  line-height: 1.5rem;
  font-weight: 400;
}

.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  margin-left: 0.25rem;
  margin-right: 0.25rem;
  border-width: 0;
  border-color: rgb(225,227,234);
  border-style: solid;
  border-bottom-width: 2px;
  width: auto;
}

@media screen and (min-width:40em) {
  .emotion-0 {
    margin-left: 2.25rem;
    margin-right: 2.25rem;
  }
}

@media screen and (min-width:52em) {
  .emotion-0 {
    margin-left: 4.5rem;
    margin-right: 4.5rem;
  }
}

<div
  className="emotion-1"
>
  <hr
    aria-orientation="horizontal"
    className="emotion-0"
    height={null}
    width="auto"
  />
</div>
`;

exports[`Separator it should set responsive vertical margins 1`] = `
.emotion-1 {
  color: rgb(18,28,45);
  font-size: 0.875rem;
  font-family: Inter,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen-Sans,Ubuntu,Cantarell,'Helvetica Neue',sans-serif;
  line-height: 1.5rem;
  font-weight: 400;
}

.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  margin-bottom: 0.25rem;
  margin-top: 0.25rem;
  border-width: 0;
  border-color: rgb(225,227,234);
  border-style: solid;
  border-left-width: 2px;
  height: auto;
}

@media screen and (min-width:40em) {
  .emotion-0 {
    margin-bottom: 2.25rem;
    margin-top: 2.25rem;
  }
}

@media screen and (min-width:52em) {
  .emotion-0 {
    margin-bottom: 4.5rem;
    margin-top: 4.5rem;
  }
}

<div
  className="emotion-1"
>
  <hr
    aria-orientation="vertical"
    className="emotion-0"
    height="auto"
    width={null}
  />
</div>
`;

exports[`Separator it should set vertical margins 1`] = `
.emotion-1 {
  color: rgb(18,28,45);
  font-size: 0.875rem;
  font-family: Inter,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen-Sans,Ubuntu,Cantarell,'Helvetica Neue',sans-serif;
  line-height: 1.5rem;
  font-weight: 400;
}

.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  margin-bottom: 2.25rem;
  margin-top: 2.25rem;
  border-width: 0;
  border-color: rgb(225,227,234);
  border-style: solid;
  border-left-width: 2px;
  height: auto;
}

<div
  className="emotion-1"
>
  <hr
    aria-orientation="vertical"
    className="emotion-0"
    height="auto"
    width={null}
  />
</div>
`;
